system/jhtree/jhinplace.hpp

```diff
@@ -0,0 +1,252 @@
+    virtual size32_t getSizeAt(unsigned int num) const override;
+    virtual offset_t getFPosAt(unsigned int num) const override;
+};
+
+class jhtree_decl CJHInplaceLeafNode : public CJHInplaceTreeNode
+{
+public:
+    virtual bool fetchPayload(unsigned int num, char *dest) const override;
+    virtual size32_t getSizeAt(unsigned int num) const override;
+    virtual offset_t getFPosAt(unsigned int num) const override;
+};
+
+class jhtree_decl CInplaceWriteNode : public CWriteNode
+{
+public:
+    CInplaceWriteNode(offset_t fpos, CKeyHdr *keyHdr, bool isLeafNode);
+
+    virtual const void *getLastKeyValue() const override { return lastKeyValue.get(); }
+    virtual unsigned __int64 getLastSequence() const override { return lastSequence; }
+
+    void saveLastKey(const void *data, size32_t size, unsigned __int64 sequence);
+
+protected:
+    MemoryAttr lastKeyValue;
+    unsigned __int64 lastSequence = 0;
+    size32_t keyCompareLen = 0;
+};
+
+class jhtree_decl CInplaceBranchWriteNode : public CInplaceWriteNode
+{
+public:
+    CInplaceBranchWriteNode(offset_t fpos, CKeyHdr *keyHdr, KeyBuildContext & _ctx);
+
+    virtual bool add(offset_t pos, const void *data, size32_t size, unsigned __int64 sequence) override;
+    virtual void write(IFileIOStream *, CRC32 *crc) override;
+
+protected:
+    unsigned getDataSize();
+
+protected:
+    KeyBuildContext & ctx;
+    PartialMatchBuilder builder;
+    Unsigned64Array positions;
+    unsigned nodeSize;
+    bool scaleFposByNodeSize = true;
+};
+
+class jhtree_decl CInplaceLeafWriteNode : public CInplaceWriteNode
+{
+public:
+    CInplaceLeafWriteNode(offset_t fpos, CKeyHdr *keyHdr, KeyBuildContext & _ctx);
+
+    virtual bool add(offset_t pos, const void *data, size32_t size, unsigned __int64 sequence) override;
+    virtual void write(IFileIOStream *, CRC32 *crc) override;
+
+protected:
+    unsigned getDataSize();
+
+protected:
+    KeyBuildContext & ctx;
+    PartialMatchBuilder builder;
+    KeyCompressor lzwcomp;
+    MemoryBuffer uncompressed;      // Much better if these could be shared by all nodes => refactor
+    MemoryAttr compressed;
+    UnsignedArray payloadLengths;
+    Unsigned64Array positions;
+    __uint64 minPosition = 0;
+    __uint64 maxPosition = 0;
+    unsigned nodeSize;
+    size32_t keyLen = 0;
+    bool scaleFposByNodeSize = false;
+    bool isVariable = false;
+    bool rowCompression = false;
+};
+
+class InplaceIndexCompressor : public CInterfaceOf<IIndexCompressor>
+{
+public:
+    InplaceIndexCompressor(size32_t keyedSize, IHThorIndexWriteArg * helper);
+    virtual ~InplaceIndexCompressor();
+
+    virtual const char *queryName() const override { return "inplace"; }
+    virtual CWriteNode *createNode(offset_t _fpos, CKeyHdr *_keyHdr, bool isLeafNode) const override
+    {
+        if (isLeafNode)
+            return new CInplaceLeafWriteNode(_fpos, _keyHdr, const_cast<KeyBuildContext &>(ctx));
```

Comment:
cast seems unnecessary/confusing

Reply:
changed the member so it is mutable - much cleaner